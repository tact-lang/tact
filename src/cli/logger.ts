export const CliLogger = () => {
    let hadErrors = false;

    const log = (message: string) => {
        hadErrors = true;
        console.log(message);
    };

    return {
        log,
        hadErrors: () => hadErrors,
    };
};
