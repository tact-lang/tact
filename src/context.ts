export class CompilerContext {

    readonly shared: { [key: symbol]: object } = {};

    constructor(args: { shared: { [key: symbol]: object } } = { shared: {} }) {
        this.shared = args.shared;
        Object.freeze(this.shared);
        Object.freeze(this);
    }

    addShared = <T>(store: symbol, key: string | number, value: T) => {
        let sh: { [key: string]: T } = {};
        if (this.shared[store]) {
            sh = { ...this.shared[store] };
        }
        sh[key] = value;
        return new CompilerContext({ shared: { ...this.shared, [store]: sh } });
    }
}

export function createContextStore<T>() {
    const symbol = Symbol();
    return {
        get(ctx: CompilerContext, key: string | number) {
            if (!ctx.shared[symbol]) {
                return null;
            }
            const m = ctx.shared[symbol] as { [key: string | number]: T };
            if (m[key]) {
                return m[key];
            } else {
                return null;
            }
        },
        all(ctx: CompilerContext): { [key: string | number]: T } {
            if (!ctx.shared[symbol]) {
                return {} as { [key: string | number]: T };
            }
            const m = ctx.shared[symbol] as { [key: string | number]: T };
            return m;
        },
        set(ctx: CompilerContext, key: string | number, v: T) {
            return ctx.addShared(symbol, key, v);
        }
    }
}