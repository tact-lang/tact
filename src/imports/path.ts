import { throwInternalCompilerError } from "../error/errors";
import { repeat } from "../utils/array";

// Witness tag. Do not use, do not export.
const pathTag = Symbol("path");

/**
 * Safe relative path
 */
export type RelativePath = {
    /**
     * Number of "../" in front of path
     */
    readonly stepsUp: number;
    /**
     * /-separated strings that go after optional ../
     */
    readonly segments: readonly string[];
    /**
     * Proof that path was created by RelativePath constructor
     */
    readonly [pathTag]: true;
};

/**
 * Constructor for relative paths
 */
const RelativePath = (
    stepsUp: number,
    segments: readonly string[],
): RelativePath => {
    if (stepsUp < 0) {
        throwInternalCompilerError("Negative number of ../ in path");
    }
    const result: RelativePath = {
        [pathTag]: true,
        stepsUp,
        segments: Object.freeze(segments),
    };
    return Object.freeze(result);
};

/**
 * Convert raw string with relative POSIX path into RelativePath
 */
export const fromString = (raw: string): RelativePath => {
    return raw.split("/").map(parseSegment).reduce(appendPath, emptyPath);
};

/**
 * Convert RelativePath to string.
 */
export const asString = ({ stepsUp, segments }: RelativePath): string => {
    return [...repeat("..", stepsUp), ...segments].join("/");
};

/**
 * Empty path, equivalent to "."
 */
export const emptyPath = RelativePath(0, []);

/**
 * Combine two relative paths
 */
const appendPath = (left: RelativePath, right: RelativePath): RelativePath => {
    const delta = right.stepsUp - left.segments.length;
    return RelativePath(left.stepsUp + Math.max(0, delta), [
        ...left.segments.slice(0, Math.max(0, -delta)),
        ...right.segments,
    ]);
};

const parseSegment = (segment: string): RelativePath => {
    if (segment === "..") {
        return RelativePath(1, []);
    }
    if (segment === "." || segment === "") {
        return emptyPath;
    }
    return RelativePath(0, [segment]);
};
