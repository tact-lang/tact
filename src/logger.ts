export type TactLogger = {
    log: (message: string) => void;
    error: (message: string) => void;
};

export const consoleLogger: TactLogger = {
    log: console.log,
    error: console.error,
};
