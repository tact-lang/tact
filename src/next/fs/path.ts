/**
 * Safe relative path
 */
export type RelativePath = {
    /**
     * Number of "../" in front of path
     */
    readonly stepsUp: number;
    /**
     * /-separated strings that go after optional ../
     */
    readonly segments: readonly string[];
};
