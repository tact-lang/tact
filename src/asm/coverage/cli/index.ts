import { Cell } from "@ton/core";
import { readFileSync, writeFileSync } from "node:fs";
import { generateHtml } from "@/asm/coverage/html";
import * as path from "node:path";
import * as fs from "node:fs";
import { collectAsmCoverage, collectFuncCoverage } from "@/asm/coverage/index";
import { CoverageSummary } from "@/asm/coverage/data";

const USAGE =
    "Usage: coverage <boc-file-path> <log-file-path> [<func-source-path> <func-mapping-path>]";

const main = () => {
    const args = process.argv.slice(2);

    const bocFilePath = args.at(0);
    const logPath = args.at(1);
    const funcSources = args.at(2);
    const funcMappingPath = args.at(3);
    const outputDir = "./";

    if (bocFilePath === undefined || logPath === undefined) {
        console.log(USAGE);
        process.exit(1);
    }

    console.log("âŒ› Generating coverage reports...");

    const cell = Cell.fromBoc(readFileSync(bocFilePath))[0];
    if (!cell) {
        console.error("Cannot parse BoC");
        process.exit(1);
    }

    const logs = readFileSync(logPath, "utf8");

    if (funcSources !== undefined && funcMappingPath !== undefined) {
        const { lines, summary } = collectFuncCoverage(
            cell,
            logs,
            funcSources,
            funcMappingPath,
        );

        printSummary("func", summary);

        if (!fs.existsSync(outputDir)) {
            fs.mkdirSync(outputDir, { recursive: true });
        }

        const combinedReportPath = path.join(outputDir, "coverage-func.html");
        writeFileSync(combinedReportPath, generateHtml(lines));
        console.log(`\nâœ… Report generated: ${combinedReportPath}`);
    }

    {
        const { lines, summary } = collectAsmCoverage(cell, logs);

        printSummary("asm", summary);

        if (!fs.existsSync(outputDir)) {
            fs.mkdirSync(outputDir, { recursive: true });
        }

        const combinedReportPath = path.join(outputDir, "coverage-asm.html");
        writeFileSync(combinedReportPath, generateHtml(lines));
        console.log(`\nâœ… Report generated: ${combinedReportPath}`);
    }
};

const printSummary = (kind: "func" | "asm", summary: CoverageSummary) => {
    const forKind = kind === "func" ? "FUNC CODE" : "ASSEMBLY CODE";
    const linesKind = kind === "func" ? "Lines" : "Instructions";

    console.log(`\nðŸ“Š COVERAGE SUMMARY FOR ${forKind}`);
    console.log("==================");
    console.log(`Coverage: ${summary.coveragePercentage.toFixed(2)}%`);
    console.log(`${linesKind}: ${summary.coveredLines}/${summary.totalLines}`);
    console.log(`Total Gas Used: ${summary.totalGas}`);
    console.log(`Executed ${linesKind}: ${summary.totalHits}`);
};

main();
