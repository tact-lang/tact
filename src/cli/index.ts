import { readFileSync } from "fs";
import { dirname, join } from "path";
import { execFileSync } from "child_process";
import { z, ZodError } from "zod";
import { createNodeFileSystem } from "../vfs/createNodeFileSystem";
import { parseAndEvalExpression } from "../optimizer/interpreter";
import { showValue } from "../types/types";
import { Config, ConfigProject, parseConfig } from "../config/parseConfig";
import { ArgParser, GetParserResult } from "./arg-parser";
import { CliErrors } from "./error-schema";
import { CliLogger } from "./logger";
import { ArgConsumer } from "./arg-consumer";
import { VirtualFileSystem } from "../vfs/VirtualFileSystem";
import { entries } from "../utils/tricks";
import { stdlibPath } from "../stdlib/path";
import { Logger, LogLevel } from "../context/logger";
import { build } from "../pipeline/build";
import { TactErrorCollection } from "../error/errors";

export const main = async () => {
    const L = CliLogger();
    const E = CliErrors(L.log);

    try {
        const argv = process.argv.slice(2);
        await processArgs(E, argv);
    } catch (e) {
        E.internal(e);
    }

    if (L.hadErrors()) {
        // https://nodejs.org/docs/v20.12.1/api/process.html#exit-codes
        process.exit(30);
    }
};

const processArgs = async (E: CliErrors, argv: string[]) => {
    const P = ArgParser(E);
    const getArgs = ArgSchema(P);

    const match = getArgs(argv);
    if (match.kind === 'ok') {
        const A = ArgConsumer(match.value);

        await parseArgs(E, A);
    } else {
        showHelp();
    }
};

const ArgSchema = (P: ArgParser) => {
    return P.tokenizer
        .add(P.string('config', 'c', 'CONFIG'))
        .add(P.string('project', 'p', 'NAME'))
        .add(P.boolean('quiet', 'q'))
        .add(P.boolean('withDecompilation', undefined))
        .add(P.boolean('func', undefined))
        .add(P.boolean('check', undefined))
        .add(P.string('eval', 'e', 'EXPRESSION'))
        .add(P.boolean('version', 'v'))
        .add(P.boolean('help', 'h'))
        .add(P.immediate)
        .end;
};

type Args = ArgConsumer<GetParserResult<ReturnType<typeof ArgSchema>>>;

const parseArgs = async (E: CliErrors, A: Args) => {
    if (A.single('help') && noUnknownParams(A)) {
        showHelp();
        return;
    }

    if (A.single('version') && noUnknownParams(A)) {
        showVersion();
        return;
    }

    const expression = A.single('eval');
    if (expression && noUnknownParams(A)) {
        evaluate(expression);
        return;
    }

    const configPath = A.single('config');
    if (configPath) {
        const F = createNodeFileSystem(dirname(configPath));
        if (!F.exists(configPath)) {
            E.configNotFound(configPath);
            return;
        }
        const configText = F.readFile(configPath).toString('utf-8');
        const config = parseConfigSafe(E, configPath, configText);
        if (!config) {
            return;
        }
        await compile(A, E, F, config);
    }

    const filePath = A.single('immediate');
    if (filePath) {
        const F = createNodeFileSystem(dirname(filePath));
        const config = createSingleFileConfig(filePath);
        await compile(A, E, F, config);
        return;
    }

    if (noUnknownParams(A)) {
        showHelp();
    }
};

const parseConfigSafe = (E: CliErrors, configPath:string, configText: string): Config | undefined => {
    try {
        return parseConfig(configText);
    } catch (e) {
        if (!(e instanceof ZodError)) {
            throw e;
        }
        E.configError(configPath, e.toString());
        return;
    }
};

export const createSingleFileConfig = (fileName: string): Config => ({
    projects: [
        {
            name: fileName,
            path: ensureExtension(fileName),
            output: './',
            options: {
                debug: true,
                external: true,
                ipfsAbiGetter: false,
                interfacesGetter: false,
            },
            mode: "full",
        },
    ],
});

const ensureExtension = (path: string): string => {
    return path.endsWith('.tact') ? path : `${path}.tact`;
};

const compile = async (A: Args, E: CliErrors, F: VirtualFileSystem, rawConfig: Config) => {
    const config = filterConfig(E, rawConfig, A.multiple('project') ?? []);

    if (!config) {
        return;
    }

    const suppressLog = A.single('quiet') ?? false;
    const logger = new Logger(suppressLog ? LogLevel.NONE : LogLevel.INFO);

    const flags = entries({
        checkOnly: A.single('check'),
        funcOnly: A.single('func'),
        fullWithDecompilation: A.single('withDecompilation'),
    });
    const setFlags = flags.filter(([, value]) => value);
    if (setFlags.length > 1) {
        E.incompatibleFlags();
    }
    const mode = flags.find(([, value]) => value)?.[0];
    const options: ExtraOptions = {};
    if (mode) {
        options.mode = mode;
    }
    setConfigOptions(config, options);

    const stdlib = createNodeFileSystem(stdlibPath, false); // Improves developer experience

    if (noUnknownParams(A)) {
        // TODO: all flags on the cli should take precedence over flags in the config
        // Make a nice model for it instead of the current mess
        // Consider making overwrites right here or something.
        await run({
            logger,
            config,
            project: F,
            stdlib,
        });
    }
};

export async function run(args: {
    logger: Logger;
    config: Config;
    project: VirtualFileSystem;
    stdlib: VirtualFileSystem;
}) {
    // Resolve projects
    const projects = args.config.projects;

    // Compile
    let success = true;
    let errorMessages: TactErrorCollection[] = [];
    
    for (const config of projects) {
        args.logger.info(`ðŸ’¼ Compiling project ${config.name} ...`);

        const built = await build({
            config,
            project: args.project,
            stdlib: args.stdlib,
            logger: args.logger,
        });
        success = success && built.ok;
        if (!built.ok && built.error.length > 0) {
            errorMessages = [...errorMessages, ...built.error];
        }
    }
    return { ok: success, error: errorMessages };
}

const filterConfig = (E: CliErrors, config: Config, projectNames: string[]): Config | undefined => {
    if (projectNames.length === 0) {
        return config;
    }

    // Check that all project names are valid
    for (const name of projectNames) {
        if (!config.projects.find((v) => v.name === name)) {
            E.noSuchProject(name);
            return;
        }
    }

    // Filter by names
    return {
        ...config,
        projects: config.projects.filter((v) => projectNames.includes(v.name)),
    };
};

type ExtraOptions = Pick<ConfigProject, 'mode'>

const setConfigOptions = (config: Config, options: ExtraOptions): void => {
    for (const project of config.projects) {
        Object.assign(project, options);
    }
};

const noUnknownParams = (args: Args): boolean => {
    if (args.isEmpty()) {
        return true;
    }

    showHelp();
    return false;
};

const showHelp = () => {
    console.log(`Usage
    $ tact [...flags] (--config CONFIG | FILE)

Flags
    -c, --config CONFIG         Specify path to config file (tact.config.json)
    -p, --project ...names      Build only the specified project name(s) from the config file
    -q, --quiet                 Suppress compiler log output
    --with-decompilation        Full compilation followed by decompilation of produced binary code
    --func                      Output intermediate FunC code and exit
    --check                     Perform syntax and type checking, then exit
    -e, --eval EXPRESSION       Evaluate a Tact expression and exit
    -v, --version               Print Tact compiler version and exit
    -h, --help                  Display this text and exit

Examples
    $ tact --version
    ${getVersion()}

Learn more about Tact:        https://docs.tact-lang.org
Join Telegram group:          https://t.me/tactlang
Follow X/Twitter account:     https://twitter.com/tact_language`);
};

const showVersion = () => {
    console.log(getVersion());
    // if working inside a git repository
    // also print the current git commit hash
    try {
        const gitCommit = execFileSync("git", ["rev-parse", "HEAD"], {
            encoding: "utf8",
            stdio: ["ignore", "pipe", "ignore"],
        }).trim();
        console.log(`git commit: ${gitCommit}`);
    } finally {
        process.exit(0);
    }
};

const evaluate = (expression: string) => {
    const result = parseAndEvalExpression(expression);
    switch (result.kind) {
        case "ok":
            console.log(showValue(result.value));
            process.exit(0);
            break;
        case "error": {
            console.log(result.message);
            process.exit(30);
        }
    }
};

const getVersion = () => {
    const packageSchema = z.object({
        version: z.string(),
    });
    
    const packageJsonPath = join(__dirname, "package.json");
    
    const pkg = packageSchema.parse(JSON.parse(readFileSync(packageJsonPath, "utf-8")));

    return pkg.version;
};
