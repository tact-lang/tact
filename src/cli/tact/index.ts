import { readFileSync } from "fs";
import { basename, dirname, join, normalize, resolve } from "path";
import { execFileSync } from "child_process";
import { z, ZodError } from "zod";
import { createNodeFileSystem } from "../../vfs/createNodeFileSystem";
import { createVirtualFileSystem } from "../../vfs/createVirtualFileSystem";
import { parseAndEvalExpression } from "../../optimizer/interpreter";
import { showValue } from "../../types/types";
import { Config, ConfigProject, parseConfig } from "../../config/parseConfig";
import { ArgParser, GetParserResult } from "../arg-parser";
import { CliErrors } from "./error-schema";
import { CliLogger } from "../logger";
import { ArgConsumer } from "../arg-consumer";
import { VirtualFileSystem } from "../../vfs/VirtualFileSystem";
import { entries } from "../../utils/tricks";
import { Logger, LogLevel } from "../../context/logger";
import { build } from "../../pipeline/build";
import { TactErrorCollection } from "../../error/errors";
import files from "../../stdlib/stdlib";
import { cwd } from "process";

export const main = async () => {
    const Log = CliLogger();
    const Errors = CliErrors(Log.log);

    try {
        const argv = process.argv.slice(2);
        await processArgs(Errors, argv);
    } catch (e) {
        Errors.unexpected(e);
    }

    if (Log.hadErrors()) {
        // https://nodejs.org/docs/v20.12.1/api/process.html#exit-codes
        process.exit(30);
    }
};

const processArgs = async (Errors: CliErrors, argv: string[]) => {
    const Parser = ArgParser(Errors);
    const getArgs = ArgSchema(Parser);

    const match = getArgs(argv);
    if (match.kind === "ok") {
        const Args = ArgConsumer(Errors, match.value);

        await parseArgs(Errors, Args);
    } else {
        showHelp();
    }
};

const ArgSchema = (Parser: ArgParser) => {
    return Parser.tokenizer
        .add(Parser.string("config", "c", "CONFIG"))
        .add(Parser.string("project", "p", "NAME"))
        .add(Parser.boolean("quiet", "q"))
        .add(Parser.boolean("with-decompilation", undefined))
        .add(Parser.boolean("func", undefined))
        .add(Parser.boolean("check", undefined))
        .add(Parser.string("eval", "e", "EXPRESSION"))
        .add(Parser.boolean("version", "v"))
        .add(Parser.boolean("help", "h"))
        .add(Parser.immediate).end;
};

const showHelp = () => {
    console.log(`Usage
$ tact [...flags] (--config CONFIG | FILE)

Flags
  -c, --config CONFIG         Specify path to config file (tact.config.json)
  -p, --project ...names      Build only the specified project name(s) from the config file
  -q, --quiet                 Suppress compiler log output
  --with-decompilation        Full compilation followed by decompilation of produced binary code
  --func                      Output intermediate FunC code and exit
  --check                     Perform syntax and type checking, then exit
  -e, --eval EXPRESSION       Evaluate a Tact expression and exit
  -v, --version               Print Tact compiler version and exit
  -h, --help                  Display this text and exit

Examples
  $ tact --version
  ${getVersion()}

Learn more about Tact:        https://docs.tact-lang.org
Join Telegram group:          https://t.me/tactlang
Follow X/Twitter account:     https://twitter.com/tact_language`);
};

type Args = ArgConsumer<GetParserResult<ReturnType<typeof ArgSchema>>>;

const parseArgs = async (Errors: CliErrors, Args: Args) => {
    if (Args.single("help")) {
        if (noUnknownParams(Errors, Args)) {
            showHelp();
        }
        return;
    }

    if (Args.single("version")) {
        if (noUnknownParams(Errors, Args)) {
            showVersion();
        }
        return;
    }

    const expression = Args.single("eval");
    if (expression) {
        if (noUnknownParams(Errors, Args)) {
            evaluate(expression);
        }
        return;
    }

    const configPath = Args.single("config");
    if (configPath) {
        const normalizedConfigPath = normalize(resolve(cwd(), configPath));
        const normalizedDirPath = normalize(
            resolve(cwd(), dirname(configPath)),
        );
        const Fs = createNodeFileSystem(normalizedDirPath, false);
        if (!Fs.exists(normalizedConfigPath)) {
            Errors.configNotFound(configPath);
            return;
        }
        const configText = Fs.readFile(normalizedConfigPath).toString("utf-8");
        const config = parseConfigSafe(
            Errors,
            normalizedConfigPath,
            configText,
        );
        if (!config) {
            return;
        }
        await compile(Args, Errors, Fs, config);
        return;
    }

    const filePath = Args.single("immediate");
    if (filePath) {
        const normalizedPath = resolve(cwd(), dirname(filePath));
        const Fs = createNodeFileSystem(normalizedPath, false);
        const config = createSingleFileConfig(basename(filePath));
        await compile(Args, Errors, Fs, config);
        return;
    }

    if (noUnknownParams(Errors, Args)) {
        showHelp();
    }
};

const parseConfigSafe = (
    Errors: CliErrors,
    configPath: string,
    configText: string,
): Config | undefined => {
    try {
        return parseConfig(configText);
    } catch (e) {
        if (!(e instanceof ZodError)) {
            throw e;
        }
        Errors.configError(configPath, e.toString());
        return;
    }
};

export const createSingleFileConfig = (fileName: string): Config => ({
    projects: [
        {
            name: fileName,
            path: ensureExtension(fileName),
            output: "./",
            options: {
                debug: true,
                external: true,
                ipfsAbiGetter: false,
                interfacesGetter: false,
            },
            mode: "full",
        },
    ],
});

const ensureExtension = (path: string): string => {
    return path.endsWith(".tact") ? path : `${path}.tact`;
};

const compile = async (
    Args: Args,
    Errors: CliErrors,
    Fs: VirtualFileSystem,
    rawConfig: Config,
) => {
    const config = filterConfig(
        Errors,
        rawConfig,
        Args.multiple("project") ?? [],
    );

    if (!config) {
        return;
    }

    const suppressLog = Args.single("quiet") ?? false;
    const logger = new Logger(suppressLog ? LogLevel.NONE : LogLevel.INFO);

    const flags = entries({
        checkOnly: Args.single("check"),
        funcOnly: Args.single("func"),
        fullWithDecompilation: Args.single("with-decompilation"),
    });
    const setFlags = flags.filter(([, value]) => value);
    if (setFlags.length > 1) {
        Errors.incompatibleFlags();
    }
    const mode = flags.find(([, value]) => value)?.[0];
    const options: ExtraOptions = {};
    if (mode) {
        options.mode = mode;
    }
    setConfigOptions(config, options);

    const stdlib = createVirtualFileSystem("@stdlib", files);

    if (noUnknownParams(Errors, Args)) {
        // TODO: all flags on the cli should take precedence over flags in the config
        // Make a nice model for it instead of the current mess
        // Consider making overwrites right here or something.
        const result = await run({
            logger,
            config,
            project: Fs,
            stdlib,
        });
        if (!result.ok) {
            Errors.setHadErrors();
        }
    }
};

export async function run(args: {
    logger: Logger;
    config: Config;
    project: VirtualFileSystem;
    stdlib: VirtualFileSystem;
}) {
    // Resolve projects
    const projects = args.config.projects;

    // Compile
    let success = true;
    let errorMessages: TactErrorCollection[] = [];

    for (const config of projects) {
        args.logger.info(`ðŸ’¼ Compiling project ${config.name} ...`);

        const built = await build({
            config,
            project: args.project,
            stdlib: args.stdlib,
            logger: args.logger,
        });
        success = success && built.ok;
        if (!built.ok && built.error.length > 0) {
            errorMessages = [...errorMessages, ...built.error];
        }
    }
    return { ok: success, error: errorMessages };
}

const filterConfig = (
    Errors: CliErrors,
    config: Config,
    projectNames: string[],
): Config | undefined => {
    if (projectNames.length === 0) {
        return config;
    }

    // Check that all project names are valid
    for (const name of projectNames) {
        if (!config.projects.find((v) => v.name === name)) {
            Errors.noSuchProject(name);
            return;
        }
    }

    // Filter by names
    return {
        ...config,
        projects: config.projects.filter((v) => projectNames.includes(v.name)),
    };
};

type ExtraOptions = Pick<ConfigProject, "mode">;

const setConfigOptions = (config: Config, options: ExtraOptions): void => {
    for (const project of config.projects) {
        Object.assign(project, options);
    }
};

const noUnknownParams = (Errors: CliErrors, Args: Args): boolean => {
    const leftoverArgs = Args.leftover();

    if (leftoverArgs.length === 0) {
        return true;
    }

    for (const argument of leftoverArgs) {
        Errors.unexpectedArgument(argument);
    }
    showHelp();
    return false;
};

const showVersion = () => {
    console.log(getVersion());
    // if working inside a git repository
    // also print the current git commit hash
    try {
        const gitCommit = execFileSync("git", ["rev-parse", "HEAD"], {
            encoding: "utf8",
            stdio: ["ignore", "pipe", "ignore"],
        }).trim();
        console.log(`git commit: ${gitCommit}`);
    } finally {
        process.exit(0);
    }
};

const evaluate = (expression: string) => {
    const result = parseAndEvalExpression(expression);
    switch (result.kind) {
        case "ok":
            console.log(showValue(result.value));
            process.exit(0);
            break;
        case "error": {
            console.log(result.message);
            process.exit(30);
        }
    }
};

const getVersion = () => {
    const packageSchema = z.object({
        version: z.string(),
    });

    const packageJsonPath = join(__dirname, "..", "..", "..", "package.json");

    const pkg = packageSchema.parse(
        JSON.parse(readFileSync(packageJsonPath, "utf-8")),
    );

    return pkg.version;
};
