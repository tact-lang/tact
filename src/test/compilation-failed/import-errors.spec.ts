import { join } from "path";
import { createSingleFileConfig, run } from "../../cli/tact";
import { createNodeFileSystem } from "../../vfs/createNodeFileSystem";
import { Logger, LogLevel } from "../../context/logger";
import { createVirtualFileSystem } from "../../vfs/createVirtualFileSystem";
import files from "../../stdlib/stdlib";

it("symlinks are not allowed", async () => {
    const result = await run({
        config: createSingleFileConfig(`symlink-parent.tact`, "./output"),
        logger: new Logger(LogLevel.NONE),
        project: createNodeFileSystem(join(__dirname, "contracts")),
        stdlib: createVirtualFileSystem("@stdlib", files),
    });
    expect(result.ok).toBe(false);
    const message = result.error.map((err) => err.message).join("; ");
    expect(message).toContain(
        "is a symbolic link which are not processed by Tact to forbid out-of-project-root accesses via symlinks",
    );
});
