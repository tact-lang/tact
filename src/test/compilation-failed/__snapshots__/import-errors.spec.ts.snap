// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fail on duplicate imports 1`] = `
"../../../../duplicate-import.tact:2:1: Duplicate import of '../../../../duplicate-import-helper.tact' in file '../../../../duplicate-import.tact'
  1 | import "./duplicate-import-helper";
> 2 | import "./duplicate-import-helper";
      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  3 | import "./duplicate-import-helper.tact";
"
`;
